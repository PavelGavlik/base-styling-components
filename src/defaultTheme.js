const defaultTheme = {
  scale: [
    0, 2, 4, 8, 16, 32, 64,
  ],
  textScale: [
    12, 16, 18, 24, 36, 48, 72,
  ],
  text: {
    fontFamily: 'Roboto,sans-serif',
    color: '#555',
    bold: 600,
  },
};

export type themeType = {
  scale: number[],
  tetxtScale: number[],
  text: {
    fontFamily: string,
    color: string,
    bold: 'bold' | 'normal' | 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 | 900
  },
};

export default defaultTheme;
