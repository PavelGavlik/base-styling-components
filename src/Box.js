// @flow
import React, { PropTypes } from 'react';
import defaultTheme from './defaultTheme';
import type { themeType } from './defaultTheme'; // eslint-disable-line no-duplicate-imports

export type BoxProps = {
  as?: () => React.Element<*> | string,
  style?: Object,
  m?: number | string,
  margin?: number | string,
  mh?: number | string,
  marginHorizontal?: number | string,
  mv?: number | string,
  marginVertical?: number | string,
  mb?: number | string,
  marginBottom?: number | string,
  ml?: number | string,
  marginLeft?: number | string,
  mr?: number | string,
  marginRight?: number | string,
  mt?: number | string,
  marginTop?: number | string,
  p?: number | string,
  padding?: number | string,
  ph?: number | string,
  paddingHorizontal?: number | string,
  pv?: number | string,
  paddingVertical?: number | string,
  pb?: number | string,
  paddingBottom?: number | string,
  pl?: number | string,
  paddingLeft?: number | string,
  pr?: number | string,
  paddingRight?: number | string,
  pt?: number | string,
  paddingTop?: number | string,
  height?: number | string,
  maxHeight?: number | string,
  maxWidth?: number | string,
  minHeight?: number | string,
  minWidth?: number | string,
  width?: number | string,
  bottom?: number | string,
  left?: number | string,
  right?: number | string,
  top?: number | string,
  flex?: number,
  backgroundColor?: string,
  border?: string,
  borderBottomColor?: string,
  borderBottomLeftRadius?: number,
  borderBottomRightRadius?: number,
  borderBottomWidth?: number,
  borderColor?: string,
  borderLeftColor?: string,
  borderLeftWidth?: number,
  borderRadius?: number,
  borderRightColor?: string,
  borderRightWidth?: number,
  borderStyle?: 'solid' | 'dotted' | 'dashed',
  borderTopColor?: string,
  borderTopLeftRadius?: number,
  borderTopRightRadius?: number,
  borderTopWidth?: number,
  borderWidth?: number,
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'stretch' | 'baseline',
  alignSelf?: 'auto' | 'flex-start' | 'flex-end' | 'center' | 'stretch' | 'baseline',
  flexBasis?: number | string,
  flexDirection?: 'row' | 'row-reverse' | 'column' | 'column-reverse',
  flexGrow?: number,
  flexShrink?: number,
  flexWrap?: 'wrap' | 'nowrap',
  justifyContent?: 'flex-start' | 'flex-end' | 'center' | 'space-between' | 'space-around',
  opacity?: number,
  overflow?: 'visible' | 'hidden' | 'scroll',
  position?: 'absolute' | 'relative' | 'fixed',
  zIndex?: number,
};

type BoxContext = {
  renderer: any,
  theme: themeType,
};

const computeBoxStyle = (theme = defaultTheme, {
  m,
  margin = m,
  mv = margin,
  marginVertical = mv,
  mh = margin,
  marginHorizontal = mh,
  mt = marginVertical,
  marginTop = mt,
  mb = marginVertical,
  marginBottom = mb,
  ml = marginHorizontal,
  marginLeft = ml,
  mr = marginHorizontal,
  marginRight = mr,
  p,
  padding = p,
  pv = padding,
  paddingVertical = pv,
  ph = padding,
  paddingHorizontal = ph,
  pt = paddingVertical,
  paddingTop = pt,
  pb = paddingVertical,
  paddingBottom = pb,
  pl = paddingHorizontal,
  paddingLeft = pl,
  pr = paddingHorizontal,
  paddingRight = pr,
  height,
  maxHeight,
  maxWidth,
  minHeight,
  minWidth,
  width,
  bottom,
  left,
  right,
  top,
  flex,
  backgroundColor,
  border,
  borderColor,
  borderBottomColor,
  borderLeftColor,
  borderRightColor,
  borderTopColor,
  borderRadius,
  borderBottomLeftRadius = borderRadius,
  borderBottomRightRadius = borderRadius,
  borderTopLeftRadius = borderRadius,
  borderTopRightRadius = borderRadius,
  borderWidth,
  borderBottomWidth = borderWidth,
  borderLeftWidth = borderWidth,
  borderRightWidth = borderWidth,
  borderTopWidth = borderWidth,
  borderStyle,
  alignItems,
  alignSelf,
  flexBasis,
  flexDirection,
  flexGrow,
  flexShrink,
  flexWrap,
  justifyContent,
  opacity,
  overflow,
  position,
  zIndex,

  ...props
}) => {
  let style = {
    position: 'relative',
    flexDirection: 'row',
    display: 'flex',
  };

  const maybeScaleProps = {
    marginBottom,
    marginLeft,
    marginRight,
    marginTop,
    paddingBottom,
    paddingLeft,
    paddingRight,
    paddingTop,
  };

  Object.keys(maybeScaleProps).forEach((prop) => {
    const value = maybeScaleProps[prop];
    if (typeof value === 'number') {
      style = {
        ...style,
        [prop]: ((value < theme.scale.length) ? `${theme.scale[value]}px` : `${value}px`),
      };
    } else if (value) {
      style = { ...style, [prop]: value };
    }
  });

  const maybePixelProps = {
    bottom,
    left,
    right,
    top,
    borderBottomWidth,
    borderLeftWidth,
    borderRightWidth,
    borderTopWidth,
    borderBottomLeftRadius,
    borderBottomRightRadius,
    borderTopLeftRadius,
    borderTopRightRadius,
  };

  Object.keys(maybePixelProps).forEach((prop) => {
    const value = maybePixelProps[prop];
    if (typeof value === 'number') {
      style = { ...style, [prop]: `${value}px` };
    } else if (value) {
      style = { ...style, [prop]: value };
    }
  });

  const maybePercentageProps = {
    height,
    maxHeight,
    maxWidth,
    minHeight,
    minWidth,
    width,
  };

  Object.keys(maybePercentageProps).forEach((prop) => {
    const value = maybePercentageProps[prop];
    if (typeof value === 'number') {
      style = { ...style, [prop]: ((value > 0 && value <= 1) ? `${value * 100}%` : `${value}px`) };
    } else if (value) {
      style = { ...style, [prop]: value };
    }
  });

  const otherProps = {
    flex,
    backgroundColor,
    border,
    borderColor,
    borderBottomColor,
    borderLeftColor,
    borderRightColor,
    borderTopColor,
    alignItems,
    alignSelf,
    borderStyle,
    flexBasis,
    flexDirection,
    flexGrow,
    flexShrink,
    flexWrap,
    justifyContent,
    opacity,
    overflow,
    position,
    zIndex,
  };

  Object.keys(otherProps).forEach((prop) => {
    const value = otherProps[prop];
    const isDefined = typeof value === 'number' || value;
    if (!isDefined) return;
    style = { ...style, [prop]: value };
  });

  return [style, props];
};

const Box = ({
  as,
  style,
  ...props
}: BoxProps, {
  renderer,
  theme,
}: BoxContext) => {
  const Component = as || 'div';
  const [boxStyle, restProps] = computeBoxStyle(theme, props);
  const rule = renderer.renderRule(() => ({
    ...boxStyle,
    ...style,
  }));
  return (
    <Component
      {...restProps}
      {...{ className: rule }}
    />
  );
};

Box.contextTypes = {
  renderer: PropTypes.object,
  theme: PropTypes.object,
};

export default Box;
