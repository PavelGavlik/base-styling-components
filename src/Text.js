// @flow
import React, { PropTypes } from 'react';
import type { BoxProps } from './Box';
import Box from './Box';  // eslint-disable-line no-duplicate-imports
import defaultTheme from './defaultTheme';
import type { themeType } from './defaultTheme'; // eslint-disable-line no-duplicate-imports

type TextProps = BoxProps & {
  fontFamily?: string,
  size?: number,
  fontSize?: number,
  align?: 'left' | 'right' | 'center' | 'justify',
  textAlign?: 'left' | 'right' | 'center' | 'justify',
  bold?: boolean,
  color?: string,
  decoration?: 'none' | 'underline' | 'line-through',
  textDecoration?: 'none' | 'underline' | 'line-through',
  transform?: 'none' | 'capitalize' | 'uppercase' | 'lowercase',
  textTransform?: 'none' | 'capitalize' | 'uppercase' | 'lowercase',
  italic?: boolean,
  lineHeight?: number,
};

type TextContext = {
  theme: themeType,
};

const computeTextStyle = (theme = defaultTheme, {
  fontFamily = theme.text.fontFamily,
  size,
  fontSize = size,
  align,
  textAlign = align,
  bold,
  color = theme.text.color,
  decoration,
  textDecoration = decoration,
  transform,
  textTransform = transform,
  italic,
  lineHeight,

  ...props
}) => {
  let style = {
    fontFamily,
    color,
  };

  if (lineHeight) {
    if (typeof lineHeight === 'number') {
      style = { ...style, lineHeight: `${lineHeight}px` };
    } else {
      style = { ...style, lineHeight };
    }
  }

  if (fontSize) {
    if (typeof fontSize === 'number') {
      style = {
        ...style,
        fontSize: (
          (fontSize < theme.textScale.length)
          ? `${theme.textScale[fontSize]}px`
          : `${fontSize}px`
        ),
      };
    } else {
      style = { ...style, fontSize };
    }
  }

  if (textAlign) {
    style = { ...style, textAlign };
  }

  if (bold) {
    style = { ...style, fontWeight: theme.text.bold };
  }

  if (textDecoration) {
    style = { ...style, textDecoration };
  }

  if (textTransform) {
    style = { ...style, textTransform };
  }

  if (italic) {
    style = { ...style, fontStyle: 'italic' };
  }

  return [style, props];
};

const Text = ({
  as,
  style,
  ...props
}: TextProps,
{
  theme,
}: TextContext) => {
  const [textStyle, restProps] = computeTextStyle(theme, props);
  return (
    <Box
      as={as}
      {...restProps}
      style={{
        ...textStyle,
        ...style,
      }}
    />
  );
};

Text.contextTypes = {
  theme: PropTypes.object,
};

export default Text;
