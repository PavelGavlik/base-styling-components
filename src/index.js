import Box from './Box';
import Text from './Text';
import defaultTheme from './defaultTheme';
import type { themeType } from './defaultTheme'; // eslint-disable-line no-duplicate-imports

export type { themeType };
export { default as Box } from './Box';
export { default as Text } from './Text';
export { default as defaultTheme } from './defaultTheme';

export default {
  Box,
  Text,
  defaultTheme,
};
