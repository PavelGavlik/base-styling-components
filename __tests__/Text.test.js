import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import chai from 'chai';
import { Text, defaultTheme } from '../src';
import rendererWithContext from './helpers/rendererWithContext';

const getStyleNode = () => (document.getElementById('fela-stylesheet').textContent);

test('Text renders correctly', () => {
  const tree = renderer.create(rendererWithContext(
    <Text>some text inside text component</Text>
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Text renders correctly as h1', () => {
  const tree = renderer.create(rendererWithContext(
    <Text as="h1">some h1 title</Text>
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Text renders correctly with set styling', () => {
  const tree = renderer.create(rendererWithContext(
    <Text margin={25} padding={25} color="red">some text inside text component</Text>
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should set correct default styling', () => {
  mount(rendererWithContext(
    <Text>some text inside box component</Text>
  ));
  chai.assert(getStyleNode().includes(`font-family:${defaultTheme.getFontFamily(defaultTheme.text.fontFamily)}`));
  chai.assert(getStyleNode().includes(`color:${defaultTheme.getColor(defaultTheme.text.color)}`));
});

it('should set correct custom styling', () => {
  mount(rendererWithContext(
    <Text fontFamily={'Arial'} size={30}>
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes('font-family:Arial'));
  chai.assert(getStyleNode().includes('font-size:30px'));
});

it('should set correct custom styling using style porp', () => {
  mount(rendererWithContext(
    <Text
      style={{
        fontFamily: 'Arial',
        fontSize: '30px',
      }}
    >
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes('font-family:Arial'));
  chai.assert(getStyleNode().includes('font-size:30px'));
});

it('should set correct font family', () => {
  const font = 'Arial';
  mount(rendererWithContext(
    <Text fontFamily={font}>
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes(`font-family:${font}`));
});

it('should set correct line height', () => {
  const size = 20;
  mount(rendererWithContext(
    <Text lineHeight={size}>
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes(`line-height:${size}px`));
});

it('should set correct line height using string value with units', () => {
  const size = '25px';
  mount(rendererWithContext(
    <Text lineHeight={size}>
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes(`line-height:${size}`));
});

it('should set correct font size', () => {
  const size = 20;
  mount(rendererWithContext(
    <Text size={size}>
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes(`font-size:${size}px`));
});

it('should set correct font size using string value with units', () => {
  const size = '4px';
  mount(rendererWithContext(
    <Text size={size}>
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes(`font-size:${size}`));
});

it('should set correct font size using string value without units', () => {
  const size = '20';
  mount(rendererWithContext(
    <Text size={size}>
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes(`font-size:${size}px`));
});

it('should set correct font size using scale value', () => {
  const size = 0;
  mount(rendererWithContext(
    <Text size={size}>
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes(`font-size:${defaultTheme.textScale[size]}`));
});

it('should set text as italic', () => {
  mount(rendererWithContext(
    <Text italic>
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes('font-style:italic'));
});

it('should transform text to uppercase', () => {
  mount(rendererWithContext(
    <Text textTransform="uppercase">
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes('text-transform:uppercase'));
});

it('should align text to center', () => {
  mount(rendererWithContext(
    <Text textAlign="center">
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes('text-align:center'));
});

it('should underline text', () => {
  mount(rendererWithContext(
    <Text textDecoration="underline">
      some text inside box component
    </Text>
  ));
  chai.assert(getStyleNode().includes('text-decoration:underline'));
});

it('should set correct text pseudo classes', () => {
  mount(rendererWithContext(
    <Text
      color="red"
      style={{
        ':hover': {
          color: 'green',
        },
      }}
    >
      hover buttonn
    </Text>
  ));
  chai.assert(getStyleNode().includes('color:red'));
  chai.assert(getStyleNode().includes(':hover{color:green}'));
});
