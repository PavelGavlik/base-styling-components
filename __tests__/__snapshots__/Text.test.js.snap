exports[`test Text renders correctly 1`] = `
<div
  className="a b c d e f g h i">
  some text inside text component
</div>
`;

exports[`test Text renders correctly as h1 1`] = `
<h1
  className="a b c d e f g h i">
  some h1 title
</h1>
`;

exports[`test Text renders correctly with set styling 1`] = `
<div
  className="a b c d e f g h i j k l m n o p q">
  some text inside text component
</div>
`;
